File: algorithms/python/AddTwoNumbers.py
# @author jasper
# @create 2021-09-24 11:50

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy_head = ListNode(0)
        p, q, curr = l1, l2, dummy_head
        carry = 0
        while p or q:
            x = p.val if p else 0
            y = q.val if q else 0
            add_sum = x + y + carry
            carry = int(add_sum / 10)
            curr.next = ListNode(int(add_sum % 10))
            curr = curr.next
            p = p.next if p else None
            q = q.next if q else None

        if carry > 0:
            curr.next = ListNode(carry)

        return dummy_head.next

File: algorithms/java/src/AddTwoNumbers.java
/**
 * @author jasper
 * @create 2021-09-24 11:36
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}


public class AddTwoNumbers {
    /**
     * 时间复杂度：O（max（m，n）），m 和 n 代表 l1 和 l2 的长度
     * 空间复杂度：O（max（m，n）），m 和 n 代表 l1 和 l2 的长度
     */
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode p = l1, q = l2, curr = dummyHead;
        int carry = 0;
        // carry 为10进制进位数
        while (p != null || q != null) {
            int x = (p != null) ? p.val : 0;
            int y = (q != null) ? q.val : 0;
            int sum = x + y + carry;
            carry = sum / 10;
            curr.next = new ListNode(sum % 10);
            curr = curr.next;
            if (p != null) {
                p = p.next;
            }
            if (q != null) {
                q = q.next;
            }
        }

        if (carry > 0) {
            curr.next = new ListNode(carry);
        }

        return dummyHead.next;
    }
}
